docs/tasks/administer-cluster/kms-provider.md

```diff
@@ -0,0 +1,182 @@
+
+See [Understanding the encryption at rest configuration.](/docs/tasks/administer-cluster/encrypt-data)
+
+## Implementing a KMS plugin
+
+To implement a KMS plugin, you can develop a new plugin gRPC server or enable a KMS plugin already provided by your cloud provider. You then integrate the plugin with the remote KMS and deploy it on the Kubernetes master.
+
+### Enabling the KMS supported by your cloud provider 
+Refer to your cloud provider for instructions on enabling the cloud provider-specific KMS plugin.
+
+### Developing a KMS plugin gRPC server
+You can develop a KMS plugin gRPC server using a stub file available for Go. For other languages, you use a proto file to create a stub file that you can use to develop the gRPC server code.
+
+* Using Go: Use the functions and data structures in the stub file: [service.pb.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.pb.go) to develop the gRPC server code 
+
+* Using languages other than Go: Use the protoc compiler with the proto file: [service.proto](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto) to generate a stub file for the specific language
+
+Then use the functions and data structures in the stub file to develop the server code.
+
+**Notes:**
+
+* kms plugin version: `v1beta1`
+
+In response to procedure call Version, a compatible KMS plugin should return v1beta1 as VersionResponse.version
+
+* message version: `v1beta1`
+
+All messages from KMS provider have the version field set to current version v1beta1
+
+* protocol: UNIX domain socket (`unix`)
+
+The gRPC server should listen at UNIX domain socket
+
+### Integrating a KMS plugin with the remote KMS
+The KMS plugin can communicate with the remote KMS using any protocol supported by the KMS.
+All configuration data, including authentication credentials the KMS plugin uses to communicate with the remote KMS, 
+are stored and managed by the KMS plugin independently. The KMS plugin can encode the ciphertext with additional metadata that may be required before sending it to the KMS for decryption.
+
+### Deploying the KMS plugin 
+Ensure that the KMS plugin runs on the same host(s) as the Kubernetes master(s).
+
+## Encrypting your data with the KMS provider
+To encrypt the data:
+
+1. Create a new encryption configuration file using the appropriate properties for the KMS provider:
+
+```yaml
+kind: EncryptionConfig
+apiVersion: v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - kms:
+        name: <display name of the KMS plugin>
+        endpoint: <UNIX domain socket listen address of the gRPC server (KMS plugin)>
+        cachesize: <number of data encryption keys (DEKs) to be cached in the clear>
+   - identity: {}
+```
```

Comment:
Since the syntax for this YAML is already outlined above, we may want to present a real sample here.